File: C_plus_plus/Remove_Duplicates_from_Sorted_Array.cpp
class Solution {
public:
    int removeDuplicates(int A[], int n) {
        if(n<=1) return n;
        int current = n-1,size = n;
        for(int i = n-2 ; i >= 0 ;--i){
            bool isDup = false;
            while(A[i]==A[current]&&i) {--i;isDup = true;}
            if(!i&&A[i]==A[current]) {--i;isDup = true;}
            for(int j = current+1,k = i+2 ; isDup && j < size ; ++j, ++k)
                A[k] = A[j];
            size = size - (current-i-1);
            current = i;
        }
        return size;
    }
};

File: java/Remove_Duplicates_from_Sorted_Array.java
// One pass solution
// time is O(n), space is O(1)
public class Solution {
    public int removeDuplicates(int[] nums) {
        int length = 0;
        for (int i = 0 ; i < nums.length; ++i) {
            if (i > 0 && nums[i] == nums[i-1]) continue;
            nums[length++] = nums[i]; 
        }
        return length;
    }
}
